bindings/rust/bench/src/s2n_tls.rs

```diff
@@ -0,0 +1,184 @@
+use crate::{
+    harness::{Buffer, Mode, TlsImpl},
+    read_to_bytes, CA_CERT_PATH, SERVER_CERT_CHAIN_PATH, SERVER_KEY_PATH,
+};
+use log::info;
+use s2n_tls::{
+    callbacks::VerifyHostNameCallback,
+    config::{Builder, Config},
+    connection::Connection,
+    enums::Blinding,
+    security::DEFAULT_TLS13,
+};
+use std::{
+    ffi::c_void,
+    io::{Read, Write},
+    os::raw::c_int,
+    pin::Pin,
+    task::Poll::Ready,
+};
+
+pub struct S2nTls {
+    c_to_s_buf: Pin<Box<Buffer>>, // need Pin to make sure C pointer to Buffer doesn't move
+    s_to_c_buf: Pin<Box<Buffer>>, // Buffer used in custom IO as context
+    c_config: Config,
+    s_config: Config,
+    c_conn: Connection,
+    s_conn: Connection,
+    c_handshaked: bool,
+    s_handshaked: bool,
+}
+
+/// Custom callback for verifying hostnames, need it to use s2n-tls safely
+struct HostNameHandler<'a> {
+    hostname: &'a str,
+}
+impl VerifyHostNameCallback for HostNameHandler<'_> {
+    fn verify_host_name(&self, hostname: &str) -> bool {
+        self.hostname == hostname
+    }
+}
+
+impl S2nTls {
+    /// Unsafe callback for custom IO C API
+    unsafe extern "C" fn send_cb(context: *mut c_void, data: *const u8, len: u32) -> c_int {
+        let context = &mut *(context as *mut Buffer);
+        let data = core::slice::from_raw_parts(data, len as _);
+        let ans = context.write(data).unwrap() as _;
+        ans
+    }
+
+    /// Unsafe callback for custom IO C API
+    unsafe extern "C" fn recv_cb(context: *mut c_void, data: *mut u8, len: u32) -> c_int {
+        let context = &mut *(context as *mut Buffer);
+        let mut data = core::slice::from_raw_parts_mut(data, len as _);
+        context.flush().unwrap();
+        let len = context.read(&mut data).unwrap();
+        if len == 0 {
+            info!("\t[blocking]");
```

Comment:
If we aren't having a main method (which I don't think we want) I think we can get rid of the log statements.